### ISSUE TITLE
from_raw(_mut)_buf returns slices with weird lifetimes

### ISSUE THREAD

**Issue by Tobba** [rust-lang/rust#20031]
These functions take a reference to the raw pointer you're trying to use and returns a slice with a lifetime based on the lifetime of your raw pointer. This behavior is not only extremely noisy, but leads to a mess of nasty transmutes or one-off variables when the raw pointer in question is calculated by i.e .offset().

A better approach might be letting the lifetime be inferred or letting the caller explicitly specify it.

**Comment by aturon** [comment]
cc me, cc @alexcrichton

**Comment by reem** [comment]
There is an open RFC to change the way the output lifetime is decided to instead use an explicit third parameter. That might be relevant to this (mostly based off of the examples in #20748).

**Comment by tomaka** [comment]
Examples that should be working but don't because of this issue:

Casting between pointer types: (http://is.gd/eW1KmQ)

```rust
struct Foo { data: *mut i8, len: uint }
impl Foo {
    fn as_slice(&self) -> &[i8] {
        unsafe {
            // value "self.data as *const i8" does not live long enough
            std::slice::from_raw_buf(&(self.data as *const i8), self.len)
        }
    }
}
```

Using raw pointers in iterators: (http://is.gd/UsePsX)

```rust
struct Iter<'a> { data: *const u8, marker: ContravariantLifetime<'a> }
impl<'a> Iterator for Iter<'a> {
    type Item = &'a [u8];
    fn next(&mut self) -> Option<&'a [u8]> {
        Some(unsafe {
            // cannot infer an appropriate lifetime for borrow expression due to conflicting requirements
            std::slice::from_raw_buf(&self.data, 1)
        })
    }
}
```

**Comment by mzabaluev** [comment]
@reem here's the [RFC draft](https://github.com/rust-lang/rfcs/pull/556)

**Comment by alexcrichton** [comment]
Fixed in #21926
